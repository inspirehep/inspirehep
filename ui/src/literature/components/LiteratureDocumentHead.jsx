import React from 'react';
import PropTypes from 'prop-types';
import { Map } from 'immutable';
import moment from 'moment';

import DocumentHead from '../../common/components/DocumentHead';
import { makeCompliantMetaDescription } from '../../common/utils';

const FULL_DATE_FORMAT = 'YYYY/MM/DD';

function LiteratureDocumentHead({ metadata, created }) {
  const title = metadata.getIn(['titles', 0, 'title']);
  const abstract = metadata.getIn(['abstracts', 0, 'value']);

  const onlineDate = moment(created).format(FULL_DATE_FORMAT);

  const arxiv = metadata.getIn(['arxiv_eprints', 0, 'value']);
  const doi = metadata.getIn(['dois', 0, 'value']);
  const citationPdfUrls = metadata.get('citation_pdf_urls');
  const authors = metadata.get('authors');

  const publicationInfo = metadata.getIn(['publication_info', 0], Map());
  const publicationDate = publicationInfo.get('year');
  const journalTitle = publicationInfo.get('journal_title');
  const journalVolume = publicationInfo.get('journal_volume');
  const journalIssue = publicationInfo.get('journal_issue');
  const pageStart = publicationInfo.get('page_start');
  const pageEnd = publicationInfo.get('page_end');
  return (
    // `citation_*` meta tags are used by Google Scholar
    // https://scholar.google.com/intl/en/scholar/inclusion.html#indexing
    <DocumentHead
      title={title}
      description={makeCompliantMetaDescription(abstract)}
    >
      <meta name="citation_title" content={title} />
      <meta name="citation_online_date" content={onlineDate} />

      {abstract && <meta name="citation_abstract" content={abstract} />}

      {publicationDate && (
        <meta name="citation_publication_date" content={publicationDate} />
      )}
      {arxiv && (
        <meta
          name="citation_technical_report_number"
          content={`arXiv:${arxiv}`}
        />
      )}
      {doi && <meta name="citation_doi" content={doi} />}

      {citationPdfUrls &&
        citationPdfUrls.map((link) => (
          <meta key={link} name="citation_pdf_url" content={link} />
        ))}

      {authors &&
        authors.map((author) => (
          <meta
            key={author.get('full_name')}
            name="citation_author"
            content={author.get('full_name')}
          />
        ))}

      {journalTitle && (
        <meta name="citation_journal_title" content={journalTitle} />
      )}
      {journalVolume && <meta name="citation_volume" content={journalVolume} />}
      {journalIssue && <meta name="citation_issue" content={journalIssue} />}
      {pageStart && <meta name="citation_firstpage" content={pageStart} />}
      {pageEnd && <meta name="citation_lastpage" content={pageEnd} />}
    </DocumentHead>
  );
}

LiteratureDocumentHead.propTypes = {
  metadata: PropTypes.instanceOf(Map).isRequired,
  created: PropTypes.string.isRequired,
};

export default LiteratureDocumentHead;
