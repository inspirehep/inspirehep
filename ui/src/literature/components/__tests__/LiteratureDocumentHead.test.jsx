import React from 'react';
import { shallow } from 'enzyme';
import { fromJS } from 'immutable';

import LiteratureDocumentHead from '../LiteratureDocumentHead';

describe('LiteratureDocumentHead', () => {
  it('renders with only title', () => {
    const wrapper = shallow(
      <LiteratureDocumentHead
        metadata={fromJS({
          titles: [
            { title: 'The title', subtitle: 'the subtitle which we dont care' },
            { title: 'Second title which we dont care' },
          ],
        })}
        created="2019-01-16T00:00:00+00:00"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders full literature', () => {
    const metadata = fromJS({
      abstracts: [
        { value: 'First abstract is important' },
        { value: 'Second, not so much' },
      ],
      titles: [{ title: 'Test Title' }],
      authors: [
        { full_name: 'Test, Author' },
        { full_name: 'Another, Author' },
      ],
      arxiv_eprints: [{ value: '1910.06344' }],
      dois: [{ value: '12.1234/1234567890123_1234' }],
      citation_pdf_urls: [
        'https://fulltext.cern/pdf/1',
        'https://fulltext.cern/pdf/2',
      ],
      publication_info: [
        {
          journal_title: 'Test Journal',
          journal_issue: 'test issue',
          journal_volume: '3',
          page_start: '12',
          page_end: '22',
          year: '1983',
        },
        { journal_title: 'Test Jornal 2 (which will not be used)' },
      ],
    });
    const wrapper = shallow(
      <LiteratureDocumentHead
        metadata={metadata}
        created="2019-01-16T00:00:00+00:00"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
