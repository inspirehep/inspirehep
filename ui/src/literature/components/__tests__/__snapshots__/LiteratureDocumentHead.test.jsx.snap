// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LiteratureDocumentHead renders full literature 1`] = `
<DocumentHead
  description="First abstract is important"
  title="Test Title"
>
  <meta
    content="Test Title"
    name="citation_title"
  />
  <meta
    content="2019/01/16"
    name="citation_online_date"
  />
  <meta
    content="First abstract is important"
    name="citation_abstract"
  />
  <meta
    content="1983"
    name="citation_publication_date"
  />
  <meta
    content="arXiv:1910.06344"
    name="citation_technical_report_number"
  />
  <meta
    content="12.1234/1234567890123_1234"
    name="citation_doi"
  />
  <meta
    content="https://fulltext.cern/pdf/1"
    key="https://fulltext.cern/pdf/1"
    name="citation_pdf_url"
  />
  <meta
    content="https://fulltext.cern/pdf/2"
    key="https://fulltext.cern/pdf/2"
    name="citation_pdf_url"
  />
  <meta
    content="Test, Author"
    key="Test, Author"
    name="citation_author"
  />
  <meta
    content="Another, Author"
    key="Another, Author"
    name="citation_author"
  />
  <meta
    content="Test Journal"
    name="citation_journal_title"
  />
  <meta
    content="3"
    name="citation_volume"
  />
  <meta
    content="test issue"
    name="citation_issue"
  />
  <meta
    content="12"
    name="citation_firstpage"
  />
  <meta
    content="22"
    name="citation_lastpage"
  />
</DocumentHead>
`;

exports[`LiteratureDocumentHead renders with only title 1`] = `
<DocumentHead
  description=""
  title="The title"
>
  <meta
    content="The title"
    name="citation_title"
  />
  <meta
    content="2019/01/16"
    name="citation_online_date"
  />
</DocumentHead>
`;
